lib/linguist/heuristics.yml

```diff
@@ -513,6 +513,14 @@ disambiguations:
   rules:
   - language: SubRip Text
     pattern: '^(\d{2}:\d{2}:\d{2},\d{3})\s*(-->)\s*(\d{2}:\d{2}:\d{2},\d{3})$'
+- extensions: ['.st']
+  rules:
+  - language: StringTemplate
+    pattern:
+    - '</\w+>'
+    - '\$\w+[($]'
+  - language: Smalltalk
+    pattern: '\A\s*[\[{(^"''\w\d#]|\w+\s*:=\s*\w+|class\s*>>\s*\w+|^\w+\s+\w+:|^Class\s*{|if(?:True|False):\s*\['
```

Comment:
`\w\d` can be reduced to just `\w`, since the characters matched by `\d` are considered "word" characters (and are thus matched by `\w`).

Having said that, the `\w+` sequences in this

~~~regexp
\w+\s*:=\s*\w+
~~~

should be replaced by `[_A-Za-z]\w*` or `(?!\d)\w+`:

~~~regexp
(?!\d)\w+\s*:=\s*(?!\d)\w+
~~~

Though I'm unfamiliar with Smalltalk, [a BNF on RosettaCode.org](https://rosettacode.org/wiki/Category:Smalltalk#Syntax_.28in_Pseudo_BNF.29) is pretty explicit about what the language considers to be a legal identifier:

~~~ebnf
assignment ::= <variable> ":=" <expression>
<variable> ::= (<underline>|<letter>)[<underline>|<letter>|digit]*
~~~

Reply:
\w includes numbers?! How have I gone this long without realising that ...

Reply:
Also underscores. 😁 Roughly speaking, `\w` is shorthand for `[A-Za-z0-9_]`. Modern, Unicode-savvy regex engines broaden that definition to cover the letters and digits of other scripts, too. Oniguruma, for example, [defines `\w`](https://github.com/kkos/oniguruma/blob/0a1d4fbaf2fc9fa3817e783245351a86a727059e/doc/RE#L50-L56) as any character with a `General_Category` of `Letter`, `Mark`, `Number`, or `Connector_Punctuation` (I believe Perl/PCRE's definition is more-or-less identical).

As you might've guessed, this treatment of `\w` is implementation-dependent, albeit a Unicode recommendation. It's also subject to other factors like Perl's `/a` (ASCII) modifier, whether or not the regex engine is running/compiled without Unicode support, and even the system's locale.

**Bottom-line:** When in doubt, specify ASCII "word" characters explicitly as `[A-Za-z0-9_]` or `[[:alnum:]_]`. It's usually easier (and cleaner) to just write `\w` instead and assume the input will never contain identifiers that use illegal characters (and most of the time, this is a pretty sane assumption to make).

Reply:
Changed to add `[a-zA-Z_]` before the `\w`s

> or (?!\d)\w+

Inline lookarounds don't seem to work (in JS envs at least)

Reply:
> Changed to add `[a-zA-Z_]` before the `\w`s

_Almost_ there. Just need to change the quantifier from `+` to `*`:

~~~diff
-[A-Za-z_]\w+
+[A-Za-z_]\w*
~~~

Otherwise, it'll miss names that are exactly one character in length, i.e.:

~~~smalltalk
a := "A"
b := "B"
~~~

> Inline lookarounds don't seem to work (in JS envs at least)

What JavaScript environment? Look-aheads have been supported in JS for as long as I remember (and look-behinds were added in more recent ECMAScript versions too, which is also widely-supported nowadays).

Reply:
*Inline* lookarounds - `(?!a)b=(?!c)d` matches `b=d` and fails on `b=dac` iirc.